L80-csharp-load-balancing.md

```diff
@@ -0,0 +1,259 @@
+        ICollection<string> IDictionary<string, object>.Keys { get; }
+        ICollection<object?> IDictionary<string, object>.Values { get; }
+        public void Set<TValue>(BalancerAttributesKey<TValue> key, TValue value);
+        public bool TryGetValue<TValue>(BalancerAttributesKey<TValue> key, out TValue value);
+        void ICollection<KeyValuePair<string, object>>.Add(KeyValuePair<string, object?> item);
+        void ICollection<KeyValuePair<string, object>>.Clear();
+        bool ICollection<KeyValuePair<string, object>>.Contains(KeyValuePair<string, object?> item);
+        void ICollection<KeyValuePair<string, object>>.CopyTo(KeyValuePair<string, object?>[] array, int arrayIndex);
+        bool ICollection<KeyValuePair<string, object>>.Remove(KeyValuePair<string, object?> item);
+        void IDictionary<string, object>.Add(string key, object? value);
+        bool IDictionary<string, object>.ContainsKey(string key);
+        object? IDictionary<string, object>.get_Item(string key);
+        bool IDictionary<string, object>.Remove(string key);
+        void IDictionary<string, object>.set_Item(string key, object? value);
+        bool IDictionary<string, object>.TryGetValue(string key, out object? value);
+        IEnumerator<KeyValuePair<string, object?>> IEnumerable<KeyValuePair<string, object>>.GetEnumerator();
+        IEnumerator IEnumerable.GetEnumerator();
+    }
+    public readonly struct BalancerAttributesKey<TValue> {
+        public BalancerAttributesKey(string key);
+        public string Key { get; }
+    }
+    public class BalancerState {
+        public BalancerState(ConnectivityState connectivityState, SubChannelPicker picker);
+        public ConnectivityState ConnectivityState { get; }
+        public SubChannelPicker Picker { get; }
+    }
+    public class ChannelState {
+        public ChannelState(AddressResolverResult resolverState, BalancerAttributes options);
+        public BalancerAttributes Options { get; set; }
+        public AddressResolverResult ResolverState { get; set; }
+    }
+    public class CompleteContext {
+        public CompleteContext();
+        public Exception? Error { get; set; }
+        public Metadata? Trailers { get; set; }
+    }
+    public class DnsAddressResolver : AddressResolver {
+        public DnsAddressResolver(Uri address, ILoggerFactory loggerFactory);
+        public override Task RefreshAsync(CancellationToken cancellationToken);
+        public override IDisposable Subscribe(IObserver<AddressResolverResult> observer);
+        protected override void Dispose(bool disposing);
+    }
+    public class DnsAddressResolverFactory : AddressResolverFactory {
+        public DnsAddressResolverFactory(ILoggerFactory loggerFactory);
+        public override string Name { get; }
+        public override AddressResolver Create(Uri address, AddressResolverOptions options);
+    }
+    public abstract class LoadBalancer : IDisposable {
```

Comment:
This also needs an `ExitIdle()` method, which will be used to tell the LB policy to exit IDLE state and start trying to connect.  Most policies won't implement this, but pick_first and a few others will.

The pick_first policy goes into IDLE state when the subchannel that was connected becomes disconnected, and it does not proactively reconnect until either (a) another RPC is sent on the channel or (b) the application explicitly asks the channel to reconnect by calling `channel.GetState(/*attempt_to_connect=*/true)`.  Note that in case (a), pick_first will return a picker that calls its `ExitIdle()` method when it sees the first pick request.

Reply:
Is that pick_first behavior specific to C++ or general across all implementations? I haven't seen it.

Reply:
It's not C++-specific; Java does the same thing.  But I think Go has not yet implemented this feature.

Reply:
Added. Note that I followed Java's lead here and called the method `RequestConnection()`.